### PR TITLE
Expose synthesized members in AST lookups

### PR DESCRIPTION
**Pull request by itaiferber**
**What's in this pull request?**
Addresses [SR-5215](https://bugs.swift.org/browse/SR-5215).

Some types and members are synthesized by derived protocol conformances (e.g. the CodingKeys member type or init(from:)/encode(to:) members from Decodable/Encodable conformance) — however, they are not visible in AST lookup if they have not been synthesized.

Exposes a LazyResolver callback for performing member synthesis where relevant during qualified lookups to synthesize these members on demand when needed.

An alternative approach to #10723.

### CODE REVIEW

**Comment by itaiferber** [comment]
@DougGregor Your proposed solution with `LazyResolver` is much cleaner and more generalizable; this turned out to be less of a sweeping change than I'd thought. A definite win-win here. Interested in hearing thoughts over #10723.

I also extended this to perform member lookup (instead of just member type lookup) which is helpful in resolving some subclassing issues I've been working on in a separate PR (where you wouldn't be able to override a superclass's synthesized method since it hadn't yet been synthesized during override lookup).

CC @jrose-apple

**Comment by itaiferber** [comment]
@swift-ci Please test

File: include/swift/AST/LazyResolver.h [hunk 1]
```diff
@@ -91,6 +91,9 @@ class LazyResolver {
   /// Resolve any implicitly-declared constructors within the given nominal.
   virtual void resolveImplicitConstructors(NominalTypeDecl *nominal) = 0;
 
+  /// Resolve an implicitly-generated member with the given name.
+  virtual void resolveImplicitMember(NominalTypeDecl *nominal, DeclName) = 0;
```
Comment (itaiferber) [inline include/swift/AST/LazyResolver.h]: Missing the `member` in `DeclName` member here

File: lib/AST/NameLookup.cpp [hunk 2]
```diff
@@ -1499,9 +1499,13 @@ bool DeclContext::lookupQualified(Type type,
...
 
-    // Make sure we've resolved implicit constructors, if we need them.
-    if (member.getBaseName() == ctx.Id_init && typeResolver)
-      typeResolver->resolveImplicitConstructors(current);
+    // Make sure we've resolved implicit members, if we need them.
+    if (typeResolver) {
+      if (member.getBaseName() == ctx.Id_init)
+        typeResolver->resolveImplicitConstructors(current);
+
+      typeResolver->resolveImplicitMember(current, member);
```
Comment (itaiferber) [inline lib/AST/NameLookup.cpp]: We could call `resolveImplicitConstructors` as part of this, too — didn't know if we'd prefer to do that or not.

File: lib/Sema/TypeCheckDecl.cpp [hunk 3]
```diff
@@ -8324,6 +8324,96 @@ void TypeChecker::addImplicitConstructors(NominalTypeDecl *decl) {
...
+          //
+          // If synthesizing conformance fails, this will produce diagnostics.
+          // If conformance checking was already in progress elsewhere, though,
+          // this could produce diagnostics twice.
+          //
+          // To prevent this duplication, we swallow the diagnostics if the
+          // state of the conformance is not Incomplete.
+          DiagnosticTransaction transaction(Context.Diags);
+          auto shouldSwallowDiagnostics =
+            conformance->getState() == ProtocolConformanceState::Incomplete;
```
Comment (itaiferber) [inline lib/Sema/TypeCheckDecl.cpp]: This condition should be inverted — we should swallow diagnostics in case the checking is already taking part elsewhere (i.e. `!= Incomplete`)

File: lib/AST/NameLookup.cpp (hunk 2 above)
Comment (DougGregor) [inline lib/AST/NameLookup.cpp]: I would like to eliminate `resolveImplicitConstructors` from `LazyResolver`; the `TypeChecker` version of `resolveImplicitMember` could call `TypeChecker:;resolveImplicitConstructors`.

**Review by DougGregor — APPROVED** [review]
This is so much better! The elimination of `resolveImplicitConstructors` can wait; this is ready now.

**Comment by itaiferber** [comment]
@DougGregor Awesome, thanks for the review! I'll push my intended fixes and merge. We can take care of `resolveImplicitConstructors` in a less pressing PR.

File: lib/AST/NameLookup.cpp (hunk 2 above)
Comment (jrose-apple) [inline lib/AST/NameLookup.cpp]: Is it possible for us to get in a loop here? Can checking the protocol result in us going back through lookupQualified?

Comment (itaiferber) [inline lib/AST/NameLookup.cpp]: I don't think so, as far as I can tell. I don't think we do qualified lookup anywhere in conformance checking. Investigating further now though.

**Comment by itaiferber** [comment]
@swift-ci Please test

File: lib/AST/NameLookup.cpp (hunk 2 above)
Comment (itaiferber) [inline lib/AST/NameLookup.cpp]: @jrose-apple So, it seems that it's possible to end up doing qualified lookups through performing an `UnqualifiedLookup` in `resolveWitnessViaLookup`. However, I'm having a hard time constructing a scenario in which it would be possible to enter a loop — since we only do lookups upwards/outwards (e.g. from nested types out), I don't think we could enter a loop since we can't have inheritance loops and things like that. Thoughts?

Comment (jrose-apple) [inline lib/AST/NameLookup.cpp]: Well, you're synthesizing members, right? So what happens when you ask for `init(from:)`from outside the type and it goes to look to see if you have a `CodingKeys`, and then that ends up checking to see if you have an `init(from:)`?

I mean, it's clearly *not* doing that, but it would be good to know why.

Comment (jrose-apple) [inline lib/AST/NameLookup.cpp]: Also wondering why `resolveWitnessViaLookup` uses unqualified lookup rather than qualified lookup.

Comment (itaiferber) [inline lib/AST/NameLookup.cpp]: @jrose-apple Following the protocol conformance checking is a bit hairy, but it looks like we should be in the clear. Once we've constructed the initial AST and have moved into type checking, we do breadth-first validation of members (at least, it's breadth-first for the first two passes), so types begin validating before references to these members can happen.

For a type like

```swift
struct OuterStruct {
    init() {
        innerStruct = InnerStruct.init(from: SomeDecoder(...))
    }

    let innerStruct: InnerStruct
    struct InnerStruct : Codable { let value: Int }
}
```
we first start validating `OuterStruct` and go member by member. Even though we hit `init()` first, we don't validate the body on the first pass, only that it has an interface type or that it's being validated. We hit `InnerStruct` and start validating protocol conformances before we hit the body of `init` and look up `init(from:)`.

By the time we ever end up doing a qualified lookup for `InnerStruct.init(from:)`, we've started checking the protocol conformance of `InnerStruct` to `Decodable`, so the state of the conformance is `Checking`. Once the conformance state is not `Incomplete`, we stop doing any lookups during checking.

**tl;dr:** We break recursion one the first pass, so this isn't an issue AFAICT.

Comment (itaiferber) [inline lib/AST/NameLookup.cpp]: As for unqualified vs. qualified lookup, it looks like we perform qualified lookup in the right places first so unqualified lookup on later passes works. But I don't have time to investigate further at the moment.

Comment (jrose-apple) [inline lib/AST/NameLookup.cpp]: All right. I guess if anything bad happens we'll find out about it soon enough. Please run it through the source compatibility suite as well, though!

Comment (itaiferber) [inline lib/AST/NameLookup.cpp]: @jrose-apple Okay, will do.

**Comment by itaiferber** [comment]
@swift-ci Please Test Source Compatibility

**Comment by jrose-apple** [comment]
I don't think it handles edits correctly.

@swift-ci Please test source compatibility
